Update `tryGetUTF8()` to return a `span<const char8_t>` instead of a `span<const char>`

cdumez opened the pull request:
#### 6aadfa61330f0e751cd4d05e3ec053c9977f5ea0
<pre>
Update `tryGetUTF8()` to return a `span&lt;const char8_t&gt;` instead of a `span&lt;const char&gt;`
<a href="https://bugs.webkit.org/show_bug.cgi?id=275823">https://bugs.webkit.org/show_bug.cgi?id=275823</a>

Reviewed by Darin Adler.

Update `tryGetUTF8()` to return a `span&lt;const char8_t&gt;` instead of a `span&lt;const char&gt;`,
since it returns UTF-8.

* Source/WTF/wtf/URL.cpp:
(WTF::percentEncodeCharacters):
* Source/WTF/wtf/text/CString.h:
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::utf8ForCharacters):
(WTF::StringImpl::utf8ForCharactersIntoBuffer):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::tryGetUTF8 const):
(WTF::StringImpl::tryGetUTF8ForCharacters):
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::tryGetUTF8 const):
* Source/WTF/wtf/text/WTFString.h:
(WTF::String::tryGetUTF8 const):
* Source/WebCore/dom/TextEncoder.cpp:
(WebCore::TextEncoder::encode const):
* Source/WebCore/platform/generic/KeyedEncoderGeneric.cpp:
(WebCore::KeyedEncoderGeneric::encodeString):
* Source/WebCore/workers/ScriptBuffer.cpp:
(WebCore::ScriptBuffer::append):

Canonical link: <a href="https://commits.webkit.org/280470@main">https://commits.webkit.org/280470@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bb8af07128728fa723d9978879288873e477e1c5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56708 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36036 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9182 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60323 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7151 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58835 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43659 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7341 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45945 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5022 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58738 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33877 "Found 60 new test failures: accessibility/accessibility-node-reparent.html accessibility/ancestor-computation.html accessibility/announcement-notification.html accessibility/aria-braillelabel.html accessibility/aria-brailleroledescription.html accessibility/aria-checked-mixed-value.html accessibility/aria-current-state-changed-notification.html accessibility/aria-current.html accessibility/aria-describedby-on-input.html accessibility/aria-description.html ... (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48949 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26802 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30657 "Found 60 new test failures: imported/w3c/web-platform-tests/FileAPI/Blob-methods-from-detached-frame.html imported/w3c/web-platform-tests/FileAPI/FileReader/Progress_event_bubbles_cancelable.html imported/w3c/web-platform-tests/FileAPI/FileReader/workers.html imported/w3c/web-platform-tests/FileAPI/FileReaderSync.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-array-buffer.any.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-array-buffer.any.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-constructor-dom.window.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-constructor-endings.html ... (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6280 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6154 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/49800 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/52609 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6551 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62006 "Built successfully") | 
| [❌ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/55949 "Found 1 new JSC stress test failure: stress/scoped-arguments-table-should-be-tolerant-for-oom.js.bytecode-cache (failure)") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/621 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6651 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53207 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/623 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49004 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53228 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/536 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/77710 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8436 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31865 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12876 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32950 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34035 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32696 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/text/CString.h:
```diff
@@ -64,6 +65,7 @@ class CString final {
     WTF_EXPORT_PRIVATE CString(const char*);
     WTF_EXPORT_PRIVATE CString(std::span<const char>);
     CString(std::span<const uint8_t>);
+    CString(std::span<const char8_t> characters) : CString(byteCast<uint8_t>(characters)) { }
```
Kind of irritating that this "forgets" the character encoding, but it's good for now.

darinadler commented on Source/WTF/wtf/text/StringImpl.h:
```diff
@@ -1369,19 +1369,19 @@ inline Ref<StringImpl> StringImpl::createByReplacingInCharacters(std::span<const
-        constexpr const char* emptyString = "";
-        return function(std::span(emptyString, emptyString));
+        constexpr std::array<char8_t, 1> empty { '\0' };
+        return function(std::span { empty.data(), empty.data() });
```
This seems a bit peculiar. Why do we need a `span` with a zero byte past the end? I understand that's what we had before but it seems unlikely callers rely on data outside the `span`.

cdumez replied:
Ah! I thought the same and an earlier version of my patch was returning a default constructed span. However, this introduced a lot of crashes. The issue seems to be the CString constructor from a span. It distinguishes the cases where the span is empty from the case where the same is empty and the pointer is null. It's trying to decide whether to constructor a null or an empty CString.

darinadler replied:
Yea that sounds right to me: non-null pointer but length zero needed.

darinadler replied:
Seems like we might be able to simplify the idiom slightly now that the span covers the whole empty array. Not have to pass a data pointer twice to a span constructor.

cdumez replied:
Given the bot failures, it looks like the code may rely on the null terminator after all. I'll investigate when I have some spare cycles.

darinadler commented on Source/WTF/wtf/text/StringImpl.h:
```diff
@@ -1405,33 +1405,34 @@ inline Expected<std::invoke_result_t<Func, std::span<const char>>, UTF8Conversio
-        constexpr const char* emptyString = "";
-        return function(std::span(emptyString, emptyString));
+        constexpr std::array<char8_t, 1> empty { '\0' };
+        return function(std::span { empty.data(), empty.data() });
```
This seems a bit peculiar. Why do we need a `span` with a zero byte past the end? I understand that's what we had before but it seems unlikely callers rely on data outside the `span`.

darinadler commented on Source/WTF/wtf/text/WTFString.h:
```diff
@@ -491,11 +491,11 @@ inline String String::substring(unsigned position, unsigned length) const
-        constexpr const char* emptyString = "";
-        return function(std::span(emptyString, emptyString));
+        constexpr std::array<char8_t, 1> empty { '\0' };
+        return function(std::span { empty.data(), empty.data() });
```
This seems a bit peculiar. Why do we need a `span` with a zero byte past the end? I understand that's what we had before but it seems unlikely callers rely on data outside the `span`.

cdumez replied:
I think we may not need the zero byte at the end. I suspect it is enough to have a non-null pointer inside the span so the CString constructor decides to constructor an empty string instead of a null string. Hopefully that constructor doesn't rely on the null terminator given than it's not included in the span. I'll give it a try.

darinadler commented on Source/WTF/wtf/text/StringImpl.h:
```diff
@@ -1369,19 +1369,19 @@ inline Ref<StringImpl> StringImpl::createByReplacingInCharacters(std::span<const
     if (characters.empty()) {
-        constexpr const char* emptyString = "";
-        return function(std::span(emptyString, emptyString));
+        constexpr std::array<char8_t, 1> empty { };
```
Wait, shouldn't this be `0`, not `1`?

cdumez replied:
Indeed.

darinadler replied:
Or maybe omit the size.

darinadler replied:
Or include the size 0 but omit the braces/initializer.

cdumez commented on Source/WTF/wtf/text/StringImpl.h:
```diff
@@ -1369,19 +1369,19 @@ inline Ref<StringImpl> StringImpl::createByReplacingInCharacters(std::span<const
-        constexpr const char* emptyString = "";
-        return function(std::span(emptyString, emptyString));
+        constexpr std::array<char8_t, 0> empty { };
+        return function(empty);
```
The bug is that the span constructed here from `empty` ends up with `span::data()` returning nullptr. We need to write this is a way that `span::data()` returns non null even though the span is empty.

cdumez replied:
```
    constexpr std::array<char8_t, 0> empty { };
    std::cout << "emptyData: " << empty.data() << std::endl;
```
`empty.data()` even returns nullptr in this case.

darinadler replied:
OK. That's really annoying, but at least we understand.

cdumez replied:
What would you think of the following then?
```
static constexpr char8_t empty = 0;
return std::span { &empty, 0 };
```

Do do you have a better proposal?

darinadler replied:
It's not beautiful but OK I think. It can probably be done with an empty array instead of a single byte, but not sure which is better. I'd probably put that in a helper function with a name like `nonNullEmptyUTF8Span` or something a little shorter instead of repeating it three times.

cdumez committed:
Update `tryGetUTF8()` to return a `span<const char8_t>` instead of a `span<const char>`
https://bugs.webkit.org/show_bug.cgi?id=275823

Reviewed by Darin Adler.

Update `tryGetUTF8()` to return a `span<const char8_t>` instead of a `span<const char>`,
since it returns UTF-8.

* Source/WTF/wtf/URL.cpp:
(WTF::percentEncodeCharacters):
* Source/WTF/wtf/text/CString.h:
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::utf8ForCharacters):
(WTF::StringImpl::utf8ForCharactersIntoBuffer):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::tryGetUTF8 const):
(WTF::StringImpl::tryGetUTF8ForCharacters):
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::tryGetUTF8 const):
* Source/WTF/wtf/text/WTFString.h:
(WTF::String::tryGetUTF8 const):
* Source/WebCore/dom/TextEncoder.cpp:
(WebCore::TextEncoder::encode const):
* Source/WebCore/platform/generic/KeyedEncoderGeneric.cpp:
(WebCore::KeyedEncoderGeneric::encodeString):
* Source/WebCore/workers/ScriptBuffer.cpp:
(WebCore::ScriptBuffer::append):

Canonical link: https://commits.webkit.org/280470@main
